cli: add a SIGUSR1 handler for abrupt exit

RaduBerinde opened the pull request:
This commit adds a handler for `SIGUSR1` which exits immediately (with `os.Exit`). This is intended to be used instead of `SIGKILL` when using code coverage instrumentation (which is dumped in an exit hook).

Note that `SIGUSR1` is a termination signal when unhandled, so this doesn't change much functionally.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110864)
<!-- Reviewable:end -->

RaduBerinde commented:
Friendly ping

knz reviewed (approved):
LGTM with comment nit.

> Reviewed 3 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110864#-:-Neo2TAt0W8uL_Er_VQN:b-a0qmnr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @renatolabs)

___
*[`pkg/cli/exit/codes.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110864#-Neo25oMB5kJw5Jbe-sD:-Neo25oN63C-Ml_lZKpg:b-skqqlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52cde3e5395df7881f97ebd6dfad7eb4d4c2740/pkg/cli/exit/codes.go#L66)):*
> ```Go
> 
> // Killed (137) indicates the server process was terminated with SIGUSR1. This
> // is the same exit code that SIGKILL would generate.
> ```

nit: not quite. (the parent process can distinguish the two) But good enough. Maybe rephrase "This is the same exit code as" to "Orchestration code should handle this exit code equivalently to what it would do if the process killed via SIGKILL".


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/cli/exit/codes.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110864#-Neo25oMB5kJw5Jbe-sD:-Neo3lRvBczByCKD7oWn:b-glreg7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52cde3e5395df7881f97ebd6dfad7eb4d4c2740/pkg/cli/exit/codes.go#L66)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

nit: not quite. (the parent process can distinguish the two) But good enough. Maybe rephrase "This is the same exit code as" to "Orchestration code should handle this exit code equivalently to what it would do if the process killed via SIGKILL".
</blockquote></details>

Bad phrasing on my part. SIGUSR1 normally returns code 138 . This returns the KILL exit code 137.

Maybe I should return the expected 138 code?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @renatolabs)

___
*[`pkg/cli/exit/codes.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110864#-Neo25oMB5kJw5Jbe-sD:-Neo8dsJ4DxNc24B1xcF:b-l1t251) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52cde3e5395df7881f97ebd6dfad7eb4d4c2740/pkg/cli/exit/codes.go#L66)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Bad phrasing on my part. SIGUSR1 normally returns code 138 . This returns the KILL exit code 137.

Maybe I should return the expected 138 code?
</blockquote></details>

Yes, you are right (this could be 138 instead)
and at the same time this is not what I was pointing out.

What I was pointing out is that the value returned here is not related to the value that a parent process would observe via wait/wait4/waitpid. The value returned by the program is shifted 8 bits to the left, and _then_ the signal code is added in the low order 8 bits.

So the value you are proposing here will not result in a parent process detecting that the termination was due to a signal (its WIFSIGNALED will return zero).

I don't believe it matters much in practice, but we should explain in the comment that it's not equivalent to what happens when the process terminates with a signal.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
cli: add a SIGUSR1 handler for abrupt exit

This commit adds a handler for `SIGUSR1` which exits immediately (with
`os.Exit`). This is intended to be used instead of `SIGKILL` when
using code coverage instrumentation (which is dumped in an exit hook).

Note that `SIGUSR1` is a termination signal when unhandled, so this
doesn't change much functionally.

Epic: none
Release note: None

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @renatolabs)

___
*[`pkg/cli/exit/codes.go` line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/110864#-Neo25oMB5kJw5Jbe-sD:-NerxsSh6dYXm9ts8Ujz:bnm0wy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c52cde3e5395df7881f97ebd6dfad7eb4d4c2740/pkg/cli/exit/codes.go#L66)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Yes, you are right (this could be 138 instead)
and at the same time this is not what I was pointing out.

What I was pointing out is that the value returned here is not related to the value that a parent process would observe via wait/wait4/waitpid. The value returned by the program is shifted 8 bits to the left, and _then_ the signal code is added in the low order 8 bits.

So the value you are proposing here will not result in a parent process detecting that the termination was due to a signal (its WIFSIGNALED will return zero).

I don't believe it matters much in practice, but we should explain in the comment that it's not equivalent to what happens when the process terminates with a signal.
</blockquote></details>

Makes sense.. So if we wouldn't have a handler for SIGUSR1, WIFSIGNALED would indicate that it was killed with this signal? But since we're handling the signal (we just put it on a channel), the situation is indistinguishable from us ignoring the signal and exiting for some other reason.

Updated the comment and the exit code.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110864)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

bors r+
